tree/tree/src/TTree.cxx

```diff
@@ -6864,6 +6896,9 @@ TTree* TTree::MergeTrees(TList* li, Option_t* options)
          newtree->ResetBranchAddresses();
          continue;
       }
+      else {
```

Comment:
I do not see by what mechanism this solve the refered to issues.  It sounded like this was a closer solution:
```
      // if (nentries == 0) continue;
      if (!newtree) {
         newtree = (TTree*)tree->CloneTree(nentries == 0 ? 0 : -1, options);
```

Reply:
The `newtree->MergeBranches(tree);` sounds like it adding a different features, namely, "create an output file that has the union of all the branches of all the files instead of using the structure of the first file".  In that context, `MergeBranches` (more aptly named `AddBranches` or `ImportBranches`) is missing one significant part which is to `BackFill` the branches.  As is the (new) branches now have less data/entries than the other branches and are actually mis-aligned.  (I.e. GetEntry(0) will get the proper  first entry of the original branch  but the semantic sum(number of entries in previous file)-th entry in the new branches.

Reply:
True, the BackFill feature is missing.

I'll check your suggestion in a separate PR.